sdk/core/azure-core/inc/context.hpp

```diff
@@ -0,0 +1,199 @@
+          type
+      = 0>
+  ContextValue(std::unique_ptr<DerivedFromValueBase>&& p) noexcept : active(4), p(std::move(p))
+  {
+  }
+
+  ContextValue(ContextValue&& other) : active(other.active)
+  {
+    switch (active)
+    {
+      case 1:
+        b = other.b;
+        break;
+      case 2:
+        i = other.i;
+        break;
+      case 3:
+        ::new (&s) std::string(std::move(other.s));
+        break;
+      case 4:
+        ::new (&p) std::unique_ptr<ValueBase>(std::move(other.p));
+        break;
+    }
+  }
+
+  ~ContextValue()
+  {
+    switch (active)
+    {
+      case 3:
+        s.~basic_string();
+        break;
+      case 4:
+        p.~unique_ptr<ValueBase>();
+        break;
+    }
+  }
+
+  ContextValue& operator=(const ContextValue& other) = delete;
+
+  template <class ExpectedType> const ExpectedType& get() const noexcept;
+
+  char alternative() const noexcept { return active; }
+};
+
+template <> inline const bool& ContextValue::get() const noexcept
+{
+  if (active != 1)
+  {
+    abort();
+  }
+  return b;
+}
+
+template <> inline const int& ContextValue::get() const noexcept
+{
+  if (active != 2)
+  {
+    abort();
+  }
+  return i;
+}
+
+template <> inline const std::string& ContextValue::get() const noexcept
+{
+  if (active != 3)
+  {
+    abort();
+  }
+  return s;
+}
+
+template <> inline const std::unique_ptr<ValueBase>& ContextValue::get() const noexcept
+{
+  if (active != 4)
+  {
+    abort();
+  }
+  return p;
+}
+
+class Context
+{
+public:
+  using time_point = std::chrono::system_clock::time_point;
+  using ContextValue = ContextValue;
+
+private:
+  struct ContextSharedState
+  {
+    std::shared_ptr<ContextSharedState> parent;
+    time_point cancelAt; // or something like this
+    std::string key;
+    ContextValue value;
+
+    explicit ContextSharedState() : cancelAt(time_point::max()) {}
+
+    explicit ContextSharedState(
+        const std::shared_ptr<ContextSharedState>& parent,
+        time_point cancelAt,
+        const std::string& key,
+        ContextValue&& value)
+        : parent(parent), cancelAt(cancelAt), key(key), value(std::move(value))
+    {
+    }
+  };
+
+  std::shared_ptr<ContextSharedState> impl_;
+
+  explicit Context(std::shared_ptr<ContextSharedState> impl) : impl_(std::move(impl)) {}
+
+public:
+  Context() : impl_(std::make_shared<ContextSharedState>()) {}
+
+  Context& operator=(const Context&) = default;
+
+  Context WithDeadline(time_point cancelWhen)
+  {
+    return Context{ std::make_shared<ContextSharedState>(
+        impl_, cancelWhen, std::string(), ContextValue{}) };
+  }
+
+  Context WithValue(const std::string& key, ContextValue&& value)
+  {
+    return Context{ std::make_shared<ContextSharedState>(
+        impl_, time_point::max(), key, std::move(value)) };
+  }
+
+  time_point CancelWhen();
+
+  const ContextValue& operator[](const std::string& key)
```

Comment:
noexcept?

Reply:
When are we documenting functions? You must say that if key is not found, then 'empty' is returned and the caller must be able to test against this.